middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,387 @@
+            raise PayloadLengthMismatchMiddlemanProtocolError(
+                f'Deserialized message payload length is {len(frame.signed_part_of_the_frame.payload)} '
+                f'instead of {frame.signed_part_of_the_frame.payload_length}.'
+            )
+
+    @classmethod
+    def _validate_payload_type(cls, payload_type: int):
+        if (
+            not hasattr(PayloadType, str(payload_type)) or
+            PayloadType[str(payload_type)] not in PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS
+        ):
+            raise PayloadTypeInvalidMiddlemanProtocolError(
+                f'Payload type {payload_type} is not valid or not registered.'
+            )
+
+    @classmethod
+    def _validate_request_id(cls, request_id: int) -> None:
+        if not isinstance(request_id, int):
+            raise RequestIdInvalidTypeMiddlemanProtocolError(
+                f'request_id is {type(request_id)} instead of int.'
+            )
+
+    @classmethod
+    def _validate_signature(cls, raw_signed_part_of_the_frame: bytes, signature: bytes, public_key: bytes) -> None:
+        try:
+            ecdsa_verify(
+                pubkey=public_key,
+                signature=signature,
+                message=raw_signed_part_of_the_frame,
+            )
+        except InvalidSignature:
+            raise SignatureInvalidMiddlemanProtocolError(
+                'Frame signature does not match its content.'
+            )
+
+    def serialize(self, private_key: bytes, add_frame_separator: bool=False, escape_encode: bool=False) -> bytes:
+        """ Parses payload into bytes message. """
+        serialized_payload = self._serialize_payload(self.payload)
+
+        assert isinstance(serialized_payload, bytes)
+        assert isinstance(add_frame_separator, bool)
+        assert isinstance(escape_encode, bool)
+
+        frame_format = self.get_frame_format()
+
+        # Create and build part of the frame which will be signed
+        signed_part_of_the_frame = Container(
+            request_id=self.request_id,
+            payload_type=self.payload_type,  # type: ignore  # pylint: disable=no-member
+            payload_length=len(serialized_payload),
+            payload=serialized_payload,
+        )
+        raw_signed_part_of_the_frame = frame_format.signed_part_of_the_frame.build(
+            signed_part_of_the_frame
+        )
+
+        # Create signature of part of the frame
+        frame_signature = ecdsa_sign(privkey=private_key, msghash=raw_signed_part_of_the_frame)
+
+        # Create frame with signature
+        frame = Container(
+            frame_signature=frame_signature,
+            signed_part_of_the_frame=signed_part_of_the_frame,
+        )
+        raw_frame = frame_format.build(frame)
+
+        # Encode raw frame if requested
+        if escape_encode:
+            raw_frame = self.escape_encode_raw_message(raw_frame)
```

Comment:
See my comment above about a function for escaping and inserting separators.

Escaping and adding separators is conceptually not a part of serialization of the frame. It's an implementation detail of the transport. It's on a different layer. So this is not the right place for it.

See how Golem messages works - you can get bytes of a serialized message without having anything added to it. The separator and escaping are the additional encoding that's not part of the serialized representation of the message.

Reply:
Please check my response in https://github.com/golemfactory/concent/pull/673#discussion_r207073727

> See how Golem messages works - you can get bytes of a serialized message without having anything added to it.

But here you can achieve that too, as those options are `False` by default. But for convenience you can turn them on with doing something like:
```
add_separator(
    escape_encode(
        message.serialize(...)
    )
)
```